Word View from DataBase, Code Refactoring

lucasadamski opened the pull request.

lucasadamski committed:
Added View to WordRepository and WordsContext

lucasadamski committed:
Implemented Word view model in order to send it to Word Index View

klaudiaRajs commented on SpanishWords.EntityFramework/Repositories/WordRepository.cs:
```diff
@@ -85,6 +86,29 @@ public IEnumerable<Word> GetAllWords()
+            try
+            {
+                var result = _db.vWordsStats.Where(n => n.UserId == userId).ToList();
+                result = ChangeNullAnswersToZero(result);
```
Opracuj to po stronie bazy danych, a nie generuj sztucznych metod tutaj. Dodaj defaultową wartość w widoku. 
https://stackoverflow.com/questions/4956656/sql-view-add-default-values-if-null

lucasadamski replied:
Done.

New View Script:

USE [SpanishWords]
GO

/****** Object:  View [dbo].[v_Words_Stats]    Script Date: 14-Feb-24 9:05:01 AM ******/
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

CREATE VIEW [dbo].[v_Words_Stats]
AS
SELECT        words.Id AS WordId, [Spanish], [English], [LexicalCategoryId], [GrammaticalGenderId], [UserId], [StatisticId], ISNULL([CorrectAnswers], 0) as CorrectAnswers, ISNULL([IncorrectAnswers], 0) as IncorrectAnswers, genders.Name AS GenderName, lexicals.Name AS LexicalName
FROM            [SpanishWords].[dbo].[Words] words JOIN
                         [SpanishWords].[dbo].[GrammaticalGenders] genders ON genders.id = words.GrammaticalGenderId JOIN
                         [SpanishWords].[dbo].[LexicalCategories] lexicals ON lexicals.id = words.LexicalCategoryId OUTER apply
                             (SELECT        count(Correct) AS CorrectAnswers
                               FROM            StudyEntries
                               WHERE        Correct = 1 AND StatisticId = words.StatisticId
                               GROUP BY StatisticId) AS statsCorrect OUTER apply
                             (SELECT        count(Correct) AS IncorrectAnswers
                               FROM            StudyEntries
                               WHERE        Correct = 0 AND StatisticId = words.StatisticId
                               GROUP BY StatisticId) AS statsIncorrect
GO

klaudiaRajs commented on SpanishWords.EntityFramework/Repositories/WordRepository.cs:
public IEnumerable GetAllWords() => ta funkcja nie jest nigdzie używana

klaudiaRajs commented on SpanishWords.EntityFramework/Repositories/WordRepository.cs:
```diff
@@ -85,6 +86,29 @@ public IEnumerable<Word> GetAllWords()
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(DBExceptionHelper.EF_QUERY_ERROR + DBExceptionHelper.GetErrorMessage(e.Message));
```
Błąd z tej linijki się nie chyba nigdzie nie loguje. Z logów nie szło wywnioskować, że widok nie istnieje w bazie danych.

lucasadamski replied:
Done.

lucasadamski committed:
Fixed Logging in View Read Method

lucasadamski committed:
Removed GetAllWords method - it's no longer used anywhere

lucasadamski committed:
Removed ChangeNullToZero Method, Altered The View in SQL DB

New View Script:

USE [SpanishWords]
GO

/****** Object:  View [dbo].[v_Words_Stats]    Script Date: 14-Feb-24 9:05:01 AM ******/
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

CREATE VIEW [dbo].[v_Words_Stats]
AS
SELECT        words.Id AS WordId, [Spanish], [English], [LexicalCategoryId], [GrammaticalGenderId], [UserId], [StatisticId], ISNULL([CorrectAnswers], 0) as CorrectAnswers, ISNULL([IncorrectAnswers], 0) as IncorrectAnswers, genders.Name AS GenderName, lexicals.Name AS LexicalName
FROM            [SpanishWords].[dbo].[Words] words JOIN
                         [SpanishWords].[dbo].[GrammaticalGenders] genders ON genders.id = words.GrammaticalGenderId JOIN
                         [SpanishWords].[dbo].[LexicalCategories] lexicals ON lexicals.id = words.LexicalCategoryId OUTER apply
                             (SELECT        count(Correct) AS CorrectAnswers
                               FROM            StudyEntries
                               WHERE        Correct = 1 AND StatisticId = words.StatisticId
                               GROUP BY StatisticId) AS statsCorrect OUTER apply
                             (SELECT        count(Correct) AS IncorrectAnswers
                               FROM            StudyEntries
                               WHERE        Correct = 0 AND StatisticId = words.StatisticId
                               GROUP BY StatisticId) AS statsIncorrect
GO
